src/Components/WebView/WebView/src/IpcSender.cs

```diff
@@ -29,7 +29,12 @@ public void ApplyRenderBatch(long batchId, RenderBatch renderBatch)
         {
             renderBatchWriter.Write(in renderBatch);
         }
-        var message = IpcCommon.Serialize(IpcCommon.OutgoingMessageType.RenderBatch, batchId, Convert.ToBase64String(arrayBuilder.Buffer, 0, arrayBuilder.Count));
+
+        var data = arrayBuilder.Buffer.AsSpan(0, arrayBuilder.Count).ToArray();
```

Comment:
By doing `ToArray` here, are we potentially allocating additional memory, which was avoided when we simply converted the array builder span to a base 64 string?

Reply:
If we convert the span to a base64 string, a string will be allocated too. In addition, the serializer has to intgrate the string again. 

I would argue here that the amount of memory might be higher with the string, because the string encoded as base 64 is much longer?

Reply:
Looking at how `IpcCommon.Serialize` is implemented, we're missing **so many** opportunities to avoid allocations 😄 

 * It uses a `params` array which could be a specific overload for this perf-critical case (`OutgoingMessageType`, `long`, `string-or-whatever-we-choose`)
 * It calls `OutgoingMessageType.ToString()` without using any [cached lookup](https://www.meziantou.net/caching-enum-tostring-to-improve-performance.htm) (unless the compiler/runtime now does this automatically)
 * The `Serialize` method calls `array.Prepend`, allocating yet another array
 * The `Serialize` method calls `JsonSerializer.Serialize` to build a string, and then appends that to another string via interpolation, instead of doing the serialization directly into a single output buffer
 * We allocate a new `ArrayBuilder` *and* a new `ArrayBuilderMemoryStream` *and* a new `RenderBatchWriter` every single time, all of which could probably be reused or pooled. The large blocks of data representing the batch are passing through multiple representations through this process.

Basically, we didn't attempt any optimizations here at all, which is likely a good thing until we found a justification for it (premature optimizations and all that :)). But now we arguably do have some justification.

If we're really concerned about perf for how we build the IPC message, we should consider making an end-to-end optimized version of the call. For example, `IpcCommon` could have a new method `SerializeRenderBatchMessage` that takes the original `RenderBatch` and generates the final IPC message string as a single process into a single output buffer, possibly using `String.Create` but I'm not 100% certain on the detail without working through it.

Reply:
@SteveSandersonMS  I have filed an issue to track the actual work we want to do [here](https://github.com/dotnet/maui/issues/12767)